### PR TITLE
RAINCATCH-1256 Workflow step

### PR DESCRIPTION
**Pull request by austincunningham**
## Motivation
Document the usage and implementation of the WFM story

## What?
- How to build step 
- How to integrate step 

Take into consideration that user may have multiple steps in single node module. 

- [x] Documentation

### CODE REVIEW

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 1]
```diff
@@ -0,0 +1,60 @@
+= Workflow Step
```
Comment (jstaffor) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc]: +[id='con-workflow-step-{chapter}']
 += Introducing a {WFM-RC-NameShort} Workflow _Step_

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+== Step Structure
+Steps are divided into to linked catergories the `module` and the `definition`.
```
Comment (jstaffor) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc]: Can you include a 1 liner for "linked categories" - as a reader, I don't know that this is.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+=== Module
+Modules can contain one or more angular directives for rendering steps in the portal and mobile applications.
```
Comment (jstaffor) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc:9]: Change "in the portal and mobile applications. " to "in the Client App (mobile) and the Cloud App."

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 4]
```diff
@@ -0,0 +1,60 @@
+= Workflow Step
+
+== Step Structure
+Steps are divided into to linked catergories the `module` and the `definition`.
+
+=== Module
+Modules can contain one or more angular directives for rendering steps in the portal and mobile applications. 
+{WFM-RC-NameShort} steps use Angular.js directives to build user interface features as part of workflows.
```
Comment (jstaffor) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc]: Can we get an image for this? ie: the 'step' showing in the "build user interface features"

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+The Edit section is displayed when workflow is executed and allows users to edit information typically this would be a form.
+The Edit section consists of the html template and directive.
+
+- **Preview Section (View)**
+
+The Preview section is displayed as a summary when a workflow has been completed. The Preview section
+consists of the html template and directive.
+
+=== Definition 
+Definition contains the step definition which outlines the properties such as name and decription of a step for the workflow in the portal application.
```
Comment (jstaffor) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc]: I think it is important to call out the full structure here ( if it's manageable in terms of the 'size' of it - ie: 100 items in the structure is too much)


If the json example below covers all of this, I'm happy to create a table using the details within the json object

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        "name": "Base Form",
+        "description": "base form used to for steps",
+        "templates": {
+            "form": "<base-form></base-form>",
+            "view": "<base></base>"
+        }
+    }
+]
+----
+The module and the definition are exported to be visibile when step package is imported.
```
Comment (jstaffor) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc:44]: Finding it hard to understand the full content in ^ line.....

This is what I'm thinking:

- The module and the definition are set to be visible when exported as a _Step_ package which means that when they are imported into another _Workflow_, they are visible to the user.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+----
+module.exports = {
+  ngModule: require('./angular/base'),
+  // Definition for this step that is being used in portal
+  definition: require('./definition.json')
+};
+----
+
+== Development 
+To develope your own step package it is recommend that you clone the link:https://github.com/feedhenry-raincatcher/step-template.git[step-template] repository and
```
Comment (jstaffor) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc]: When developing a custom _Step_  package, it is ...

**Review by jstaffor — CHANGES_REQUESTED** [review]

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 1 above)
Comment (wtrocki) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: More like "Building your first/Reusing template"

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 2 above)
Comment (wtrocki) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: Already covered below.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 3 above)
Comment (wtrocki) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc:9]: @jstaffor why we changing that?

**Comment by wtrocki** [comment]
@austincunningham - are we still using this tech approved labels?

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 4 above)
Comment (wtrocki) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: to build user interface features => `to build user interface`. 

@jstaffor - misunderstanding - users will build their user interface here - having image may be confusing IMHO

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 5 above)
Comment (wtrocki) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: `decription` of step sounds awesome. 
Structure is bellow.. so let's not mention that at all.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 6 above)
Comment (wtrocki) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc:44]: @jstaffor - this content is basically impossible to understand without basic knowledge of angular and our application structure. 
@austincunningham - as mentioned before please make sure that we mention that at the beginning.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 2 above)
Comment (jstaffor) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: There is no other mention of "linked catergories", "linked" or "catergories" anywhere else in this document.

According to the docs, "Edit Section (Form)" and "Preview Section (View)" are "two types of directive".

Are these the "linked catergories"?


"the `module` and the `definition`"
so the module and the definition are linked directly one cant exist without the other.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 3 above)
Comment (jstaffor) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc:9]: @wtrocki - this is more inline with other documentation. (If it's wrong to say this, no probs leaving it as is).

"The {WFM-RC-NameShort} Datasync Module is an implementation of the {WFM-RC-NameShort} Sync Framework. Datasync is made up of two applications, the Cloud App and the Client App."[1]


[1] https://github.com/feedhenry-raincatcher/raincatcher-docs/blob/master/docs/workforce-management-framework/topics/con-datasync-module.adoc

@jstaffor There is a cloud , portal and mobile application so the cloud app is not the portal if that makes sense 
image

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 4 above)
Comment (jstaffor) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: Ahh, I understand now - happy not to have an image. :+1:

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 5 above)
Comment (jstaffor) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: I'd prefer to remove the example and add useful reference material, in terms of a table which details the Step definition.
The structure below is an example and as a reader, I was wondering what the full structure is after looking at it.

**Review by jstaffor — DISMISSED** [review]

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 3 above)
Comment (jstaffor) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc:9]: That makes much more sense now @wtrocki :+1: 

Given the original:

"Modules can contain one or more angular directives for rendering steps in the portal and mobile applications."

Which of the following is most applicable:

1) "Modules can contain one or more angular directives for rendering steps in the Portal and the Client App (mobile)."
2) "Modules can contain one or more angular directives for rendering steps in the Portal, the Client App (mobile) and the Cloud App." (think local development where I have both running locally (I could be over complicating this - call it out if so))

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 2 above)
Comment (jstaffor) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: Right. I thought ""linked categories" was some sort of technical definition. So on that basis, any of the following reworks of the original sentence appeal to you:

A) "A _Step_ is divided into two categories, the `module`  category and the `definition` category and both categories are linked."

B) "A _Step_ consists of a `module` and a `definition`."

C) "A _Step_ consists of a `module` and a `definition`, and both are linked."

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 3 above)
Comment (austincunningham) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc:9]: I think the confusion here is that the feedhenry sync documentation refers to client and cloud.  Portal , mobile are what they are called in the platform , I think we should match the platforms naming conventions. So leaving as is.

Comment (wtrocki) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc:9]: @jstaffor - I think @austincunningham added some foto to your comment. 
Our names are now as above. That's are short names. 
Long names (suggestion) are posted on our main website.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 5 above)
Comment (austincunningham) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: not sure what you mean , the full structure for a single step is in the example

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 2 above)
Comment (austincunningham) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: The module and the definition can be expanded to have more that one step. so using the singular step would be incorrect.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 5 above)
Comment (jstaffor) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: Best to ignore my comment at present. I'll hold off making these types of comments until "Tech Ack" has been approved for the PR as I could be causing more confusion.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 6 above)
Comment (austincunningham) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc:44]: this is how elements are exposed in node see https://www.sitepoint.com/understanding-module-exports-exports-node-js/ for context.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 8]
```diff
@@ -0,0 +1,63 @@
+= Building your first Workflow Step/Reusing template
+
+A basic knowledge of Angular.js is requried to build a Workflow Step.
+
```
Comment (witmicko) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc]: to use Raincatcher 😉

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+== Step Structure
+Steps are divided into to linked catergories the `module` and the `definition`.
+
+=== Module
+Modules can contain one or more angular directives for rendering steps in the portal and mobile applications. 
+{WFM-RC-NameShort} steps use Angular.js directives to build user interface as part of workflows.
+Each step is divided into two types of directive:
```
Comment (witmicko) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc]: directives

File: docs/workforce-management-framework/topics/con-workflow-step.adoc [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+- **Preview Section (View)**
+
+The Preview section is displayed as a summary when a workflow has been completed. The Preview section
+consists of the html template and directive.
+
+=== Definition 
+Definition contains the step definition which outlines the properties such as name and decription of a step for the workflow in the portal application.
+Each step requires a definition so the `definition.json` file can contain a single definition or 
+an array of definitions depending on the number of steps that are included in the step package.
+
```
Comment (witmicko) [inline docs/workforce-management-framework/topics/con-workflow-step.adoc]: Each step requires a definition, the  `definition.json` file can contain a definition of a single step or an array of definitions with the size corresponding to the number of steps in the step package.
?

**Review by witmicko — COMMENTED** [review]
Looks good, couple minor comments.

File: docs/workforce-management-framework/topics/con-workflow-step.adoc (hunk 9 above)
Comment (wtrocki) [reply docs/workforce-management-framework/topics/con-workflow-step.adoc]: Each step is divided into two directives. "Types of directive" is misleading.

**Review by wtrocki — APPROVED** [review]

**Comment by wtrocki** [comment]
@jstaffor My suggestion is to merge this now as it will be part of the bigger vision that needs to be rearranged in master.

@wtrocki - this does need a review from a CCS pov. I'll keep my review purely as a CCS review.

**Comment by wtrocki** [comment]
@jstaffor Did not saw your reply as you reviewed my comment, but yeah fully agree there is no rush to push that content + it's the most important part of the documentation.

**Comment by jstaffor** [comment]
I'll focus on this PR in the morning with the aim of having it done mid day

**Comment by wtrocki** [comment]
@austincunningham This PR weren't created on the upstream repo. 
John created separate PR with the fixes on the top level so I'm going to close this one. 
https://github.com/feedhenry-raincatcher/raincatcher-docs/pull/73

**Pull request closed without merging**